### PR TITLE
[Tracer] Enable 128 bits by default

### PR DESCRIPTION
**Pull request by pierotibou**
## Summary of changes

Enable 128 bits trace id generation by default.

## Reason for change

It's the new default we should support as part of the OTEL changes.

## Implementation details

Just switch the flag. Added a helper function in TraceId.

## Test coverage

Coverage seemed good. Just changed the tests.
Also ignored `_dd.p.tid` in smoke tests.

## Other details
We add the propagation tags in CI Visibility which seems useless as they aren't handled by the backend. We could not add them in that case. But I'll do this in a different PR.

AIT-4244

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (4827)](https://github.com/DataDog/dd-trace-dotnet/tree/08ad071237613bcc79a55badb4bccdda50e1df5f)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/6c051698537075c2c4883bc78ea8f248c8e29465)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (72ms)  : 63, 81
     .   : milestone, 72,
    master - mean (70ms)  : 63, 78
     .   : milestone, 70,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (1,000ms)  : 979, 1021
     .   : milestone, 1000,
    master - mean (998ms)  : 981, 1015
     .   : milestone, 998,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (106ms)  : 103, 110
     .   : milestone, 106,
    master - mean (106ms)  : 103, 108
     .   : milestone, 106,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (693ms)  : 671, 716
     .   : milestone, 693,
    master - mean (684ms)  : 664, 705
     .   : milestone, 684,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (90ms)  : 86, 94
     .   : milestone, 90,
    master - mean (90ms)  : 87, 92
     .   : milestone, 90,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (659ms)  : 633, 684
     .   : milestone, 659,
    master - mean (660ms)  : 636, 683
     .   : milestone, 660,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.6.2) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (188ms)  : 185, 190
     .   : milestone, 188,
    master - mean (187ms)  : 185, 190
     .   : milestone, 187,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (1,100ms)  : 1083, 1117
     .   : milestone, 1100,
    master - mean (1,096ms)  : 1078, 1114
     .   : milestone, 1096,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (272ms)  : 268, 275
     .   : milestone, 272,
    master - mean (272ms)  : 267, 277
     .   : milestone, 272,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (1,050ms)  : 1023, 1077
     .   : milestone, 1050,
    master - mean (1,050ms)  : 1034, 1066
     .   : milestone, 1050,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (4827) - mean (263ms)  : 259, 266
     .   : milestone, 263,
    master - mean (263ms)  : 259, 266
     .   : milestone, 263,

    section CallTarget+Inlining+NGEN
    This PR (4827) - mean (1,020ms)  : 1000, 1040
     .   : milestone, 1020,
    master - mean (1,027ms)  : 1001, 1053
     .   : milestone, 1027,

```

**Comment by andrewlock** [comment]
## Throughput/Crank Report:zap:

Throughput results for AspNetCoreSimpleController comparing the following branches/commits:
- [This PR (4827)](https://github.com/DataDog/dd-trace-dotnet/tree/08ad071237613bcc79a55badb4bccdda50e1df5f)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/1bc8121e4019906390c70ed783fe11a898996526)
- [benchmarks/2.9.0](https://github.com/DataDog/dd-trace-dotnet/tree/89ae3f4ee7d85b824badda5873166707c9e560b4)

Cases where throughput results for the PR are worse than latest master (5% drop or greater), results are shown in **red**.

Note that these results are based on a _single_ point-in-time result for each branch. For full results, see [one](https://ddstaging.datadoghq.com/dashboard/fnz-afi-c2i/apm-net-crank) of the [many](https://ddstaging.datadoghq.com/dashboard/c5g-i7d-kcy/ciapp-apm-net-throughput-tests), [many](https://ddstaging.datadoghq.com/dashboard/uxh-m8j-qhi/ciapp-apm-net-throughput-tests-kevin) dashboards!

```mermaid
gantt
    title Throughput Linux x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4827) (11.153M)   : 0, 11153074
    master (11.290M)   : 0, 11290445
    benchmarks/2.9.0 (11.049M)   : 0, 11048936

    section Automatic
    This PR (4827) (7.649M)   : 0, 7649014
    master (7.699M)   : 0, 7698544
    benchmarks/2.9.0 (7.950M)   : 0, 7949791

    section Trace stats
    This PR (4827) (8.011M)   : 0, 8011483
    master (8.024M)   : 0, 8023555

    section Manual
    This PR (4827) (9.869M)   : 0, 9869197
    master (9.614M)   : 0, 9614162

    section Manual + Automatic
    This PR (4827) (7.266M)   : 0, 7266019
    master (7.281M)   : 0, 7280980

    section Version Conflict
    This PR (4827) (6.611M)   : 0, 6611323
    master (6.549M)   : 0, 6549445

```
```mermaid
gantt
    title Throughput Linux arm64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4827) (9.687M)   : 0, 9686741
    master (9.479M)   : 0, 9478674
    benchmarks/2.9.0 (9.490M)   : 0, 9489921

    section Automatic
    This PR (4827) (6.636M)   : 0, 6635925
    master (6.655M)   : 0, 6654533

    section Trace stats
    This PR (4827) (6.883M)   : 0, 6882572
    master (6.956M)   : 0, 6956041

    section Manual
    This PR (4827) (8.241M)   : 0, 8240785
    master (8.272M)   : 0, 8271591

    section Manual + Automatic
    This PR (4827) (6.084M)   : 0, 6084159
    master (6.323M)   : 0, 6323247

    section Version Conflict
    This PR (4827) (5.674M)   : 0, 5673988
    master (5.688M)   : 0, 5688212

```
```mermaid
gantt
    title Throughput Windows x64 (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    This PR (4827) (9.665M)   : 0, 9664726
    master (9.542M)   : 0, 9542238
    benchmarks/2.9.0 (9.740M)   : 0, 9739913

    section Automatic
    This PR (4827) (6.712M)   : 0, 6712250
    master (6.888M)   : 0, 6888184
    benchmarks/2.9.0 (7.236M)   : 0, 7236066

    section Trace stats
    This PR (4827) (7.011M)   : 0, 7011256
    master (7.223M)   : 0, 7223001

    section Manual
    This PR (4827) (8.431M)   : 0, 8431181
    master (8.752M)   : 0, 8751559

    section Manual + Automatic
    This PR (4827) (6.669M)   : 0, 6669173
    master (6.730M)   : 0, 6729743

    section Version Conflict
    This PR (4827) (5.822M)   : crit ,0, 5821750
    master (6.148M)   : 0, 6148245

```
```mermaid
gantt
    title Throughput Linux x64 (ASM) (Total requests) 
    dateFormat  X
    axisFormat %s
    section Baseline
    master (7.365M)   : 0, 7364969
    benchmarks/2.9.0 (7.964M)   : 0, 7964040

    section No attack
    master (2.121M)   : 0, 2120871
    benchmarks/2.9.0 (3.235M)   : 0, 3234651

    section Attack
    master (1.677M)   : 0, 1677422
    benchmarks/2.9.0 (2.544M)   : 0, 2544176

    section Blocking
    master (3.406M)   : 0, 3405896

    section IAST default
    master (6.795M)   : 0, 6795160

    section IAST full
    master (6.053M)   : 0, 6053152

    section Base vuln
    master (0.956M)   : 0, 956193

    section IAST vuln
    master (0.941M)   : 0, 940764

```

File: tracer/src/Datadog.Trace/Configuration/TracerSettings.cs [hunk 1]
```diff
@@ -353,7 +353,7 @@ internal TracerSettings(IConfigurationSource? source, IConfigurationTelemetry te
 
             TraceId128BitGenerationEnabled = config
                                             .WithKeys(ConfigurationKeys.FeatureFlags.TraceId128BitGenerationEnabled)
-                                            .AsBool(false);
+                                            .AsBool(true);
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Configuration/TracerSettings.cs:356]: I assume we don't have any back-compat concerns with this right?

File: tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs [hunk 2]
```diff
@@ -262,7 +262,7 @@ public void FixTraceIdTag(TraceId traceId)
                 // (parse the hex string and compare ulongs to avoid allocating another string)
                 if (tagValue == null || !HexString.TryParseUInt64(tagValue, out var currentValue) || currentValue != traceId.Upper)
                 {
-                    SetTag(Trace.Tags.Propagated.TraceIdUpper, HexString.ToHexString(traceId.Upper));
+                    SetTag(Trace.Tags.Propagated.TraceIdUpper, traceId.ToUpperHexString());
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs:265]: I'm not a fan of this at all 😛  `traceId.ToUpperHexString()` and `ToHexString(traceId.Upper)` mean very different things to me. The former means "convert `traceId` to an uppercase hex string". The latter means "convert the upper bytes to a hex string".

Also, the `ToUpperHexString()` methods seems to do exactly the same as `HexString.ToHexString(Upper)` so I don't see why it's necessary 😉 

```suggestion
                    SetTag(Trace.Tags.Propagated.TraceIdUpper, HexString.ToHexString(traceId.Upper));
```

File: tracer/src/Datadog.Trace/TraceId.cs [hunk 3]
```diff
@@ -84,6 +84,14 @@ public int CompareTo(object? obj)
...
     public override int GetHashCode() => HashCode.Combine(Lower, Upper);
 
+    [Pure]
+    public string ToUpperHexString()
+    {
+        return this == Zero ?
+                   "0000000000000000" :
+                   HexString.ToHexString(Upper);
+    }
+
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/TraceId.cs:94]: :kill-it-with-fire:
```suggestion
```

File: tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs (hunk 2 above)
Comment (pierotibou) [reply tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs:265]: It does the same yes, it's a copy paste. I like to have the logic of what we want to do with those traceids in the trace id function. But I don't mind reverting if you' re really against (I mean my prio is to deliver, not refacto :)) 
I understand the name could be better. `UpperToHexString` would have been better. 
So no go and I revert? or `UpperToHexString` is ok?

Comment (andrewlock) [reply tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs:265]: I mean, I'd prefer to revert, but see how others feel 😉

Comment (pierotibou) [reply tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs]: Done in 5879a4e6f33dc6e8fc21d3bf2e03cdf01eeb20d1

File: tracer/src/Datadog.Trace/TraceId.cs (hunk 3 above)
Comment (pierotibou) [reply tracer/src/Datadog.Trace/TraceId.cs]: Done in 5879a4e6f33dc6e8fc21d3bf2e03cdf01eeb20d1

File: tracer/src/Datadog.Trace/Configuration/TracerSettings.cs (hunk 1 above)
Comment (pierotibou) [reply tracer/src/Datadog.Trace/Configuration/TracerSettings.cs:356]: This is what I understood. We can anyway bring @zacharycmontoya or @lucaspimentel in the loop to double confirm.

Comment (zacharycmontoya) [reply tracer/src/Datadog.Trace/Configuration/TracerSettings.cs:356]: Correct, there are no back-compat concerns according to our OpenTelemetry team

**Review by andrewlock — APPROVED** [review]

**Review by robertpi — APPROVED** [review]
image
LGTM WCPGW

**Review by andrewlock — APPROVED** [review]

**Pull request merged**